Metrics

ranlu opened the pull request:
Collect metrics data using statsd-exporter/prometheus/grafana. The grafana dashboard can be access through `https://[bootstrap_node]/grafana`, the setup loosely follows https://databand.ai/blog/everyday-data-engineering-monitoring-airflow-with-prometheus-statsd-and-grafana/. But I separated seuron related metrics from airflow metrics and created customized dashboard.

ranlu committed:
Add metrics services

ranlu committed:
Pass statsd parameters

ranlu committed:
Firewall rule for grafana

ranlu committed:
Add statsd parameters to the parameter files

ranlu committed:
Add metadata for igneous tasks

Right now I only use it for metrics collection, can extend it for other
purposes in the future

ranlu committed:
Send igneous/custom metrics to statsd

There were some old code send metrics using the airflow Stats object,
switch to raw statsd interface so we can customize the name

ranlu committed:
Do not expose prometheus port

prometheus pulls data from exporters, only grafana needs to access
it directly, they are both inside the docker network, no need to open
the external port

nicholasturner1 commented on deploy/docker-compose-CeleryExecutor.yml:
```diff
@@ -60,6 +60,39 @@ services:
+            - "9125:9125/udp"
+
+    prometheus:
+        image: ranlu/prometheus
```
I think it would be helpful to store all of the Dockerfiles you use in this repo for easier reference.

ranlu replied:
Sure, I want to put them in a separate repo though, since the changes to these images are not so tightly related seuron, (for example, add metrics in corgie or synaptor to measure performance and visualize them in grafana). `seuron_deployment` would be a appropriate place if it is not taken already

nicholasturner1 commented on dags/igneous_and_cloudvolume.py:
```diff
@@ -79,6 +79,13 @@ def chunk_tasks(tasks, chunk_size):
      for i in range(0, len(tasks), chunk_size):
          yield tasks[i:i + chunk_size]
 
+def tasks_with_metadata(run_name, tasks):
```
I'm curious about your git log message - what other applications do you have in mind for this?

ranlu replied:
For the hard coded tasks I do not have any immediate idea, for customized tasks we can use the metadata to setup the queue, how to ramp up the cluster, etc. For example we can merge custom cpu tasks and custom gpu tasks and pick the appropriate one, can even mix them together in one script.

nicholasturner1 reviewed (approved):
Looks good :+1:

ranlu committed:
Only try to get metadata from a dict

In case the function returns a list of tasks, set the metadata to an
empty dict.

ranlu commented:
Found a bug when testing with igneous tasks, the code fails when task script returns a list. Fixed it with the last commit, @nicholasturner1, can you take a quick look before I merge the pr?

nicholasturner1 commented on dags/igneous_and_cloudvolume.py:
```diff
@@ -158,8 +158,9 @@ def inner(*args, **kwargs):
             try:
                 ret = create_tasks(*args, **kwargs)
-                metadata = ret.get('metadata', {})
+                metadata = {}
```
This makes sense to me.
